### ISSUE TITLE
"Subroutine * redefined" because of cycling inheritance + use self in parent

### ISSUE THREAD

**Issue by nugged** [jberger/Mojo-TypeModel#2]
If I have:

```perl
package MyProj::Model;
use Mojo::Base 'Mojo::TypeModel';
use MyProj::Model::Users;
```
and:
```perl
package MyProj::Model::Users;
use Mojo::Base 'MyProj::Model';

sub whatever ...
```
any sub-module subroutine from MyProj::Model::_ANY_ will have this "Subroutine redefined" warning because this child-level module cyclically use'd in parent module so it inherits own subs again.

Removing that line `use MyProj::Model::Users;` out from main `MyProj::Model` and 'use' it in another place is enough but makes pieces of code slightly scattered, yes.

**Comment by jberger** [comment]
Would it help if you load the types by using `require` rather than `use`, ie

```perl
package MyProj::Model;
use Mojo::Base 'Mojo::TypeModel';
require MyProj::Model::Users;
```

If so then I should change the documentation suggesting the pattern you are following. Interesting that I haven't noticed anything.

**Comment by nugged** [comment]
nope, still "redefined", I also slightly confused because it is not exporting anything out,
so it's somehow tricky even for explain for me why it get's redefined.

I thought that Mojo::Base parent do something with exporting,
tried also `use MyProj::Model::Users ();` -> no, it is still warns.

How that cyclical thing happens, some "Mojo::Base" fuzzy magic imports, so that warnings happen?

I also tried now:
```perl
package MyProj::Model::Users;
use strict; use warnings;
# use Mojo::Base -strict;
use base 'MyProj::Model';
```
to test without 'Mojo::Base': same "redefined". 
I feel I am doing unnecessary voodoo dancing here :) because issues lie somewhere else.

Anyway, for now, moved that out to the upper module:
```perl
use MyProj::Model ();
use MyProj::Model::Users ();
```

(btw Perl v5.24.2 here under perlbrew, just to note).

**Comment by nugged** [comment]
(edited comment above, added more tries)

> Interesting that I haven't noticed anything.

That's why it makes me confused twice because it's a very simple piece of code, but maybe I have some own additional "something" ... will dig more now, re-build example, etc.

I tested with:

```shell
lib:$>perl -c MyProj/Model/Users.pm
Subroutine test redefined at MyProj/Model/Users.pm line 5.
MyProj/Model/Users.pm syntax OK
```

**Comment by jberger** [comment]
Ok I can get the warnings when I `perl -c` on the library. I hadn't done that before. When I load the library into a mojo application then everything seems to work without warnings. I'll have to dig in more to understand and see if I can either fix it or offer something advice in the docs.

**Comment by nugged** [comment]
```shell
CarPark/lib:$>perl -c CarPark/Model/User.pm
Subroutine add_user redefined at CarPark/Model/User.pm line 8.
Subroutine set_password redefined at CarPark/Model/User.pm line 24.
Subroutine check_password redefined at CarPark/Model/User.pm line 31.
Subroutine add_roles redefined at CarPark/Model/User.pm line 37.
Subroutine remove_roles redefined at CarPark/Model/User.pm line 44.
Subroutine check_role redefined at CarPark/Model/User.pm line 51.
Subroutine _user redefined at CarPark/Model/User.pm line 57.
CarPark/Model/User.pm syntax OK
```

Checked on your CarPark... Ah, I see you got it also.

**Comment by jberger** [comment]
Right but now try just using the application: `$ perl script/carpark eval -V 'app->model->user'`

**Comment by nugged** [comment]
Yes, it won't cry when not syntax-checked alone. My Mojo engine quiet too.
Some Perl tricks?  

>  if I can either fix it or offer something advice in the docs.

just wished to assist for more robust tool / future users. Worried just in case if that might bring something more dangerous if Perl interprets it in cycle way like leakage or whatever else we even can't imagine :).

**Comment by jberger** [comment]
I think I'm going to phone-a-friend. @haarg got any thoughts?

**Comment by jberger** [comment]
BTW, I can fix it by deferring the model loading to the beginning of the run phase:

```perl                                                                                  
use CarPark::Model::Door;                                                           
use CarPark::Model::GPIO;                                                           
use CarPark::Model::User;                                                            
```

becomes

```perl
INIT {                                                                                    
  require CarPark::Model::Door;                                                           
  require CarPark::Model::GPIO;                                                           
  require CarPark::Model::User;                                                           
}  
```

But I don't love deferring it that way (though it should be ok, I think) and that is more cognitive overhead for users.

**Comment by nugged** [comment]
> But I don't love deferring

I agree that hacks-a-like solutions are not so glossy :). And that requires a looot of explanation for Juniors. ¯\\\_(ツ)\_/¯
